Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -213,14 +213,19 @@ private class TopSitesCollectionView: UICollectionView {
 
 private class TopSitesLayout: UICollectionViewLayout {
     private var thumbnailRows: Int {
-        return max(2, Int((self.collectionView?.frame.height ?? self.thumbnailHeight) / self.thumbnailHeight))
+        return max(2, Int((self.height ?? self.thumbnailHeight) / self.thumbnailHeight))
     }
 
     private var thumbnailCols = 2
     private var thumbnailCount: Int {
         return thumbnailRows * thumbnailCols
     }
-    private var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
+
+    private lazy var width: CGFloat = { return self.collectionView?.frame.width ?? 0 }()
+
+    private lazy var height: CGFloat? = {
+        return self.collectionView?.frame.height
+        }()
```

Comment:
Ah I think I know the problem with the extra row at the bottom. It looks like viewDidLoad calls height/width and at that point the collectionView's frame hasn't been updated by autolayout yet so it reports a height of 667 which will result in 5 rows. Since these are lazy, height/width will only evaluate once and cache that result. Instead of being lazy, these should be computed properties so we don't cache the height/width values